### PR TITLE
Lock-based thread safety

### PR DESCRIPTION
**Pull request by crusaderky**
- Partially closes https://github.com/dask/distributed/pull/7686
- The whole lockless thread safety design from #82 and #90 was extremely brittle. This PR swaps it for an industry-standard RLock based design.
- Full thread safety for all classes (except Zip and LMDB), with much less caveats than before
- Follow-up: #99

### CODE REVIEW

File: zict/buffer.py [hunk 1]
```diff
@@ -94,48 +105,74 @@ def fast_to_slow(self, key: KT, value: VT) -> None:
...
-            del self.slow[key]
-        except KeyError:
-            pass
-        # This may trigger an eviction from fast to slow of older keys.
-        # If the weight is individually greater than n, then key/value will be stored
-        # into self.slow instead (see LRU.__setitem__).
+        with self._lock:
+            discard(self.slow, key)
+            if key in self._cancel_restore:
+                self._cancel_restore[key] = True
```
Comment (crusaderky) [inline zict/buffer.py:150]: Annoyingly I can't just call
```python
self.set_noevict(key, value)
self.fast.evict_until_below_capacity()
```
due to the bespoke exception handling in ``LRU.__setitem__``.
I'll clean this up once distributed has switched to async mode.

File: zict/lru.py [hunk 2]
```diff
@@ -54,38 +55,50 @@ class LRU(ZictBase[KT, VT]):
...
     closed: bool
-    total_weight: Accumulator
+    total_weight: float
+    _cancel_evict: dict[KT, bool]
 
     def __init__(
         self,
         n: float,
         d: MutableMapping[KT, VT],
+        *,
```
Comment (crusaderky) [inline zict/lru.py:70]: Very small API breakage; I doubt anybody will mind

File: zict/tests/test_lru.py [hunk 3]
```diff
@@ -283,7 +281,7 @@ def test_getitem_is_threasafe():
 
     def f(_):
         barrier.wait()
-        for _ in range(5_000_000):
+        for _ in range(500_000):
             assert lru["x"] == 1
```
Comment (crusaderky) [inline zict/tests/test_lru.py:285]: This test has become an order of magnitude slower.
I'm not too concerned - this is just during continuous collisions from 2+ threads; if only a thread is involved, the overhead of the RLock is nanoseconds.

**Comment by crusaderky** [comment]
@milesgranger this is ready for review.
I appreciate it's _a lot_ to digest. Take your time, and if you need me to walk you through it, ping me.

File: zict/tests/utils_test.py [hunk 4]
```diff
@@ -1,9 +1,21 @@
...
 from collections.abc import ItemsView, KeysView, MutableMapping, ValuesView
+from concurrent.futures import ThreadPoolExecutor
 
 import pytest
 
+from zict.common import ZictBase
+
+# How many times to repeat non-deterministic stress tests.
+# You may set it as high as 50 if you wish to run in CI.
+REPEAT_STRESS_TESTS = 1
```
Comment (crusaderky) [inline zict/tests/utils_test.py:17]: I ran it 100 times just before review; all green.

File: zict/cache.py [hunk 5]
```diff
@@ -44,44 +42,75 @@ class Cache(ZictBase[KT, VT]):
...
+    @locked
+    def __setitem__(self, key: KT, value: VT) -> None:
+        # If the item was already in cache and data.__setitem__ fails, e.g. because
+        # it's a File and the disk is full, make sure that the cache is invalidated.
+        discard(self.cache, key)
+        gen = self._gen
+        gen += 1
+        self._last_updated[key] = self._gen = gen
+
+        with self.unlock():
```
Comment (milesgranger) [inline zict/cache.py:88]: Can we comment here why this benefits from being unlocked? It seems like some of the second condition below can be removed if this were left locked. Assume there is a reason for the trade-off though; to allow another thread to make progress? If so, I'm curious if it's worth the tradeoff for added complexity of checking / correcting state (second conditional call to `discard(self.cache, key)`) in that same second conditional.

File: zict/file.py [hunk 6]
```diff
@@ -123,13 +128,9 @@ def __setitem__(
...
     ) -> None:
-        try:
-            del self[key]
-        except KeyError:
-            pass
-
+        self.discard(key)
         fn = self._safe_key(key)
-        with open(os.path.join(self.directory, fn), "wb") as fh:
+        with open(os.path.join(self.directory, fn), "wb") as fh, self.unlock():
```
Comment (milesgranger) [inline zict/file.py:133]: Curious why writing files is considered 'safe' to release the lock but deleting it is not (see __delitem__)? Seems mutation of files ought to (probably) be locked, or consistently unlocked and left to user to worry about.

File: zict/sieve.py [hunk 7]
```diff
@@ -36,63 +37,83 @@ class Sieve(ZictBase[KT, VT], Generic[KT, VT, MKT]):
...
         self.key_to_mapping[key] = mapping
+        self.gen += 1
+        gen = self.gen
+
+        with self.unlock():
+            mapping[key] = value
 
+        if gen != self.gen and self.key_to_mapping.get(key) is not mapping:
+            # Multithreaded race condition
+            discard(mapping, key)
```
Comment (milesgranger) [inline zict/sieve.py:77]: I suppose this is a similar clarification; is the tradeoff  worth potentially corrupting the expected state and trying to rectify it?

**Review by milesgranger — APPROVED** [review]
Looks pretty darn good to me; just some nits and clarifications about implementation. Specifically why we still need to account for potential state mgmt caused from the temporary releases of the locks and if it's worth releasing those locks during those points. 

However it seems well tested and thought out, so feel free to just comment and move along with approval; I have nothing which warrants holding you up here. :)

File: zict/common.py [hunk 8]
```diff
@@ -66,3 +103,27 @@ def flush(*z: Any) -> None:
...
+    """
+    try:
+        del m[key]
+    except KeyError:
+        pass
+
+
+def locked(func: Callable[P, VT]) -> Callable[P, VT]:
+    """Decorator for a method of ZictBase, which wraps the whole method in a
+    mapping-global rlock.
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply zict/common.py:120]: Yes, per-instance.

File: zict/cache.py (hunk 5 above)
Comment (crusaderky) [reply zict/cache.py:88]: It's explained in the notes at the top of every class.
These methods can take arbitrarily long to run, so two threads should be able to run in unison on the same mapping while they're running:

- `Buffer.slow.__setitem__`
- `Buffer.slow.__getitem__`
- `LRU.on_evict`
- `Cache.data.__setitem__`
- `Cache.data.__getitem__`
- `File.__setitem__` (specifically fh.write / fh.writelines; not the other syscalls)
- `File.__getitem__` (specifically fh.readinto; not the other syscalls)

All other methods are expected to be _fast_, which should be intended as _it's OK for a thread that must not be busy_ (read: the Worker's event loop) _to lock waiting for them to finish_.

File: zict/file.py (hunk 6 above)
Comment (crusaderky) [reply zict/file.py:133]: We are not mutating files though. We're always creating a new one. If two threads call `__setitem__` on the same key, they'll end up writing to two different files.
I just noticed that I forgot about a race condition there though - one of the two files would remain there indefinitely, littering the hard drive. Fixed.

It's not that deleting files is unsafe - it's that `os.remove` is expected to be _fast_ (see definition above) so I chose to keep the method simple. For the record, `dict.pop` (first line of `__delete__`) is *not* thread-safe and would need to be wrapped in a lock.

**Comment by milesgranger** [comment]
Thanks for the clarifications  @crusaderky :smiley:

**Pull request merged**